internal/federation/server.go

```diff
@@ -224,7 +231,20 @@ func (s *Server) SendFederationRequest(deployment *docker.Deployment, req gomatr
 	}
 
 	httpClient := gomatrixserverlib.NewClient(gomatrixserverlib.WithTransport(&docker.RoundTripper{Deployment: deployment}))
-	return httpClient.DoRequestAndParseResponse(context.Background(), httpReq, resBody)
+
+	start := time.Now()
+	err = httpClient.DoRequestAndParseResponse(context.Background(), httpReq, resBody)
+
+	var outcome string
+	if respError, ok := err.(gomatrix.RespError); ok {
+		outcome = respError.ErrCode
+	} else if httpError, ok := err.(gomatrix.HTTPError); ok {
+		outcome = strconv.Itoa(httpError.Code)
```

Comment:
The status code alone isn't enough. You need to print the HTTP response body. Print `.Contents` - https://github.com/matrix-org/gomatrixserverlib/blob/f72a635100606c4e772a60d9028e9fe1c98850e3/client.go#L539

Reply:
(happy to be persuaded otherwise though.)

Reply:
> it's the test's responsibility to print out more detailed errors

This is a great way to get no detailed errors, because people either forget, don't know how, or can't spend the time to add such detail. Complement is designed to give **good** errors by default, and detailed errors with `COMPLEMENT_DEBUG=1` set. Please ensure the response body is logged if `COMPLEMENT_DEBUG=1` is set.